### PR TITLE
fix negation in dynamic_bitset

### PR DESCRIPTION
**Pull request by olk**

### CODE REVIEW

**Comment by mclow** [comment]
Um... what does this fix?  [I'm missing the context for this change]

**Comment by eldiener** [comment]
The fix is that negation should be on 0 as an integer, not on 0 as a Block. This fix looks correct to me.

**Comment by mclow** [comment]
> The fix is that negation should be on 0 as an integer, not on 0 as a Block.

Yes, I see that's what the code is doing.
But that's not what I was asking.

**Comment by olk** [comment]
compiling unit-tests with _gcc -W -Wall -Werror_ produces errors like:

../../../boost/dynamic_bitset/dynamic_bitset.hpp:1953:57: error: left shift of negative value [-Werror=shift-negative-value]
         block_type const mask = (~static_cast<Block>(0) << extra_bits);

File: include/boost/dynamic_bitset/dynamic_bitset.hpp [hunk 1]
```diff
@@ -1070,7 +1070,7 @@ bool dynamic_bitset<Block, Allocator>::all() const
                 return false;
             }
         }
-        block_type const mask = ~(~static_cast<Block>(0) << extra_bits);
+        block_type const mask = ~(static_cast<Block>(~0) << extra_bits);
```
Comment (mclow) [inline include/boost/dynamic_bitset/dynamic_bitset.hpp]: So much strangeness here.  You're making a value, casting it to a `Block`, and then assigning it to a `block_type`. If they were different types, you'd be doing a conversion. But they're not - so why use two different names?
Also, we know that `block_type` is an unsigned integral type, right? So why start with a signed constant `0`?
Why not just write:
     `block_type const mask = (block_type(1) << extra_bits) - 1;`
Isn't that the same value? No more `~`s, no `static_cast`.

**Comment by olk** [comment]
I've added a bug report

**Pull request closed without merging**